File: py/annoy.py
# annoy.py — Approximate Nearest Neighbors Oh Yeah
#
# Annoy builds a forest of random binary trees. Each tree recursively
# splits the dataset by picking two random points and creating a
# hyperplane (perpendicular bisector) between them. At query time,
# traverse each tree to a leaf, union the candidates, then brute-force
# rank them to find the true k nearest.
#
# Run tests with: python3 test_annoy.py

from __future__ import annotations
from dataclasses import dataclass, field
from typing import Optional
import random

from knn import euclidean_distance, classify


# A node in the random binary tree.
# Leaf nodes store point indices. Internal nodes store a hyperplane
# (normal vector + offset) and left/right children.
@dataclass
class Node:
    is_leaf: bool
    indices: list[int] = field(default_factory=list)
    left: Optional[Node] = None
    right: Optional[Node] = None
    normal: list[float] = field(default_factory=list)
    offset: float = 0.0


# ============================================================
# STEP 1: Dot product
# ============================================================
def _dot(a: list[float], b: list[float]) -> float:
    """Return the dot product of two equal-length vectors.

    a · b = Σ(a_i * b_i)
    """
    return sum([ai * bi for ai, bi in zip(a, b)])


# ============================================================
# STEP 2: Split hyperplane
# ============================================================
def _make_split_hyperplane(
    point_a: list[float],
    point_b: list[float],
) -> tuple[list[float], float]:
    """Compute the hyperplane that is the perpendicular bisector between
    point_a and point_b.

    A hyperplane in R^n is defined by: { x | n · x = d }
    where n is the normal vector and d is the offset.

    For the perpendicular bisector:
      - The normal n = b - a (points from a toward b)
      - The plane passes through the midpoint m = (a + b) / 2
      - The offset d = n · m

    Side test: a point x is on the LEFT if n · x - d <= 0

    Returns:
        (normal, offset)
    """
    normal = [bi - ai for ai, bi in zip(point_a, point_b)]
    midpoint  = [(ai + bi)/2 for ai, bi in zip(point_a, point_b)]
    offset = _dot(normal, midpoint)
    return  normal, offset


class AnnoyIndex:
    """An approximate nearest neighbor index using a forest of random
    binary trees.

    Usage:
        index = AnnoyIndex(n_trees=10, max_leaf_size=10)
        index.build(training_data)
        neighbors = index.query(query_point, k=5)

    Parameters:
        n_trees:       Number of random trees to build. More trees = better
                       recall, slower build and query.
        max_leaf_size: Stop splitting when a node has this many or fewer
                       points.
    """

    # ============================================================
    # STEP 3: Constructor
    # ============================================================
    def __init__(self, n_trees: int = 10, max_leaf_size: int = 10) -> None:
        """Store hyperparameters and initialize an empty forest."""
        self.n_trees = n_trees
        self.max_leaf_size = max_leaf_size
        self._forest: list[Node] = []
        self._training_data: list[tuple[list[float], str]] = []

    def _split_over_hyperplane(
        self,
        indices: list[int],
        hyperplane: tuple[list[float], float]
    ) -> tuple[list[int], list[int]]:
        normal, offset = hyperplane
        left: list[int] = []
        right: list[int] = []
        for idx in indices:
            point = self._training_data[idx][0]
            is_left = (_dot(normal, point) - offset) <= 0
            if is_left:
                left.append(idx)
            else:
                right.append(idx)
        return left, right


    # ============================================================
    # STEP 4: Build one tree
    # ============================================================
    def _build_tree(self, indices: list[int]) -> Node:
        """Recursively build one random binary tree over the given point
        indices.

        Base case:
            |indices| <= max_leaf_size → return a leaf.

        Recursive case:
            1. Sample two distinct points p, q from the subset.
            2. Compute the perpendicular bisector hyperplane between p and q.
            3. Partition all indices by which side of the hyperplane they
               fall on: n · x_i - d <= 0 → left, otherwise → right.
            4. If either partition is empty, return a leaf (degenerate split).
            5. Recurse on each partition.

        Returns:
            Root Node of this subtree.
        """
        if (len(indices) <= self.max_leaf_size):
            return Node(is_leaf=True, indices = indices)

        idx_a, idx_b = random.sample(indices, 2)
        point_a, point_b = self._training_data[idx_a][0], self._training_data[idx_b][0]
        hyperplane = _make_split_hyperplane(point_a, point_b)
        left_idxs, right_idxs = self._split_over_hyperplane(indices, hyperplane)
        if len(left_idxs) == 0 or len(right_idxs) == 0:
            # degenerate split
            return Node(is_leaf=True, indices = indices)
        left = self._build_tree(left_idxs)
        right = self._build_tree(right_idxs)
        normal, offset = hyperplane
        return Node(is_leaf=False, left=left, right=right, normal=normal, offset=offset)

    # ============================================================
    # STEP 5: Build the forest
    # ============================================================
    def build(self, training_data: list[tuple[list[float], str]]) -> None:
        """Build n_trees independent random trees over the full dataset.

        Each tree sees all points but makes different random splits.
        """
        self._training_data = training_data

        all_idxs = list(range(len(training_data)))
        for _ in range(self.n_trees):
            self._forest.append(self._build_tree(all_idxs))

    # ============================================================
    # STEP 6: Query one tree
    # ============================================================
    def _query_tree(self, node: Node, query: list[float]) -> list[int]:
        """Traverse a single tree to find the leaf containing the query.

        At each internal node, test which side of the hyperplane the
        query falls on: n · q - d <= 0 → go left, else → go right.

        Returns the indices stored in the reached leaf.
        """
        if (node.is_leaf):
            return node.indices
        assert node.left # Node dataclass needs better typing?
        assert node.right
        is_left = (_dot(node.normal, query) - node.offset) <= 0
        if (is_left):
            return self._query_tree(node.left, query)
        else:
            return self._query_tree(node.right, query)
        

    # ============================================================
    # STEP 7: Query the forest
    # ============================================================
    def query(self, query: list[float], k: int) -> list[int]:
        """Find approximate k nearest neighbors.

        1. Collect candidate indices from all trees (union/deduplicate).
        2. Rank candidates by true distance to query.
        3. Return the top k.
        """
        candidate_set: set[int] = set()
        for tree in self._forest:
            candidate_set.update(self._query_tree(tree, query))
        ranked = sorted(candidate_set, key=lambda i: euclidean_distance(self._training_data[i][0], query))
        return ranked[:k]



    # ============================================================
    # STEP 8: Evaluate
    # ============================================================
    def evaluate(
        self,
        test_data: list[tuple[list[float], str]],
        k: int,
    ) -> float:
        """Classification accuracy: for each test point, query the index,
        majority-vote the neighbors' labels, compare to true label.

        Returns accuracy in [0.0, 1.0].
        """
        correct = sum(1 for features, label in test_data if classify(self.query(features, k), self._training_data) == label)
        return correct / len(test_data)

File: cpp/src/annoy.cpp
#include "star_nn/annoy.h"
#include "star_nn/knn.h"

#include <algorithm>
#include <cassert>
#include <cstddef>
#include <fstream>
#include <iostream>
#include <numeric>
#include <random>
#include <ranges>
#include <span>
#include <stdexcept>
#include <unordered_set>
#include <vector>

namespace star_nn {

using std::size_t;

double dot(const Vector &a, const Vector &b) {
  return dot(std::span<const double>{a}, std::span<const double>{b});
}

double dot(std::span<const double> a, std::span<const double> b) {
  assert(a.size() == b.size());

  // a · b = Σ(a_i * b_i)
  double d = 0.0;
  for (const auto [ai, bi] : std::views::zip(a, b)) {
    d += ai * bi;
  }
  return d;
}

std::pair<Vector, double> make_split_hyperplane(const Vector &a,
                                                const Vector &b) {
  Vector normal = std::vector<double>();
  Vector midpoint = std::vector<double>();
  normal.reserve(a.size());
  midpoint.reserve(a.size());

  for (const auto [ai, bi] : std::views::zip(a, b)) {
    normal.push_back(bi - ai);
    midpoint.push_back((ai + bi) / 2.0);
  }
  const auto offset = dot(normal, midpoint);

  return {normal, offset};
}

AnnoyIndex::AnnoyIndex(std::size_t n_trees, std::size_t max_leaf_size)
    : n_trees_(n_trees), max_leaf_size_(max_leaf_size) {}

std::pair<std::vector<size_t>, std::vector<size_t>>
AnnoyIndex::split_over_hyperplane_(const std::vector<size_t> &indices,
                                   const Vector &normal, double offset) {
  auto left = std::vector<size_t>();
  auto right = std::vector<size_t>();
  left.reserve(indices.size());
  right.reserve(indices.size());

  for (const auto idx : indices) {
    const auto &point = this->training_data_[idx].features;
    const auto is_left = (dot(normal, point) - offset) <= 0;
    if (is_left)
      left.push_back(idx);
    else
      right.push_back(idx);
  }
  return {left, right};
}

std::size_t AnnoyIndex::build_tree_(const std::vector<size_t> &indices) {
  if (indices.size() <= this->max_leaf_size_) {
    std::size_t leaf_item_offset = leaf_items_.size();
    leaf_items_.insert(leaf_items_.end(), indices.begin(), indices.end());

    auto node = DiskNode{
        .left_index = -1,  // Leaf
        .right_index = -1, // Leaf
        .leaf_item_offset = leaf_item_offset,
        .leaf_item_count = indices.size(),
        .normal_offset = 0,
        .hyperplane_offset = 0.0,
    };

    nodes_.push_back(node);
    return nodes_.size() - 1;
  }

  static std::mt19937 rng(std::random_device{}());
  std::uniform_int_distribution<std::size_t> dist(0, indices.size() - 1);
  const std::size_t pos_a = dist(rng);
  std::size_t pos_b = dist(rng);
  while (pos_b == pos_a) {
    pos_b = dist(rng);
  }
  const auto idx_a = indices[pos_a];
  const auto idx_b = indices[pos_b];

  const auto &point_a = training_data_[idx_a].features;
  const auto &point_b = training_data_[idx_b].features;

  const auto [normal, offset] = make_split_hyperplane(point_a, point_b);
  const auto [left_indices, right_indices] =
      split_over_hyperplane_(indices, normal, offset);

  if (left_indices.empty() || right_indices.empty()) {
    std::size_t leaf_item_offset = leaf_items_.size();
    leaf_items_.insert(leaf_items_.end(), indices.begin(), indices.end());

    auto node = DiskNode{
        .left_index = -1,  // Leaf
        .right_index = -1, // Leaf
        .leaf_item_offset = leaf_item_offset,
        .leaf_item_count = indices.size(),
        .normal_offset = 0,
        .hyperplane_offset = 0.0,
    };

    nodes_.push_back(node);
    return nodes_.size() - 1;
  }

  const auto left = build_tree_(left_indices);
  const auto right = build_tree_(right_indices);

  const auto normal_offset = normals_.size();
  normals_.insert(normals_.end(), normal.begin(), normal.end());
  auto node = DiskNode{
      .left_index = static_cast<std::int64_t>(left),
      .right_index = static_cast<std::int64_t>(right),

      .leaf_item_offset = 0,
      .leaf_item_count = 0,
      .normal_offset = normal_offset,
      .hyperplane_offset = offset,
  };

  nodes_.push_back(node);
  return nodes_.size() - 1;
}

void AnnoyIndex::build(const Dataset &training_data) {
  this->training_data_ = training_data;
  this->dims_ = training_data_[0].features.size();

  std::vector<std::size_t> indices(training_data_.size());
  std::iota(indices.begin(), indices.end(), 0);

  for (size_t i = 0; i < n_trees_; ++i) {
    std::cout << "  building tree " << (i + 1) << "/" << n_trees_ << "...\n";
    auto root = build_tree_(indices);
    forest_.push_back(root);
  }
}

template <typename T>
void write_vector(std::ofstream &out, const std::vector<T> &v) {
  out.write(reinterpret_cast<const char *>(v.data()),
            static_cast<std::streamsize>(v.size() * sizeof(T)));
}

template <typename T> void read_vector(std::ifstream &in, std::vector<T> &v) {
  in.read(reinterpret_cast<char *>(v.data()),
          static_cast<std::streamsize>(v.size() * sizeof(T)));
  if (!in) {
    throw std::runtime_error("failed to read Annoy index section");
  }
}

void AnnoyIndex::save(const std::string &path) const {
  std::ofstream out(path, std::ios::binary);

  AnnoyFileHeader header = {
      .version = 1,
      .dims = dims_,
      .n_trees = n_trees_,
      .max_leaf_size = max_leaf_size_,
      .forest_count = forest_.size(),
      .node_count = nodes_.size(),
      .leaf_item_count = leaf_items_.size(),
      .normal_count = normals_.size(),
  };

  out.write(reinterpret_cast<const char *>(&header), sizeof(header));

  write_vector(out, forest_);
  write_vector(out, nodes_);
  write_vector(out, leaf_items_);
  write_vector(out, normals_);
}
void AnnoyIndex::load(const std::string &path, const Dataset &training_data) {
  std::ifstream in(path, std::ios::binary);
  if (!in) {
    throw std::runtime_error("failed to open Annoy index file: " + path);
  }

  AnnoyFileHeader header;
  in.read(reinterpret_cast<char *>(&header), sizeof(header));
  if (!in) {
    throw std::runtime_error("failed to read Annoy index header");
  }

  if (header.version != 1) {
    throw std::runtime_error("unsupported Annoy index version");
  }

  if (training_data.empty()) {
    throw std::runtime_error("cannot load Annoy index with empty training data");
  }

  if (header.dims != training_data[0].features.size()) {
    throw std::runtime_error("Annoy index dimensions do not match training data");
  }

  dims_ = header.dims;
  n_trees_ = header.n_trees;
  max_leaf_size_ = header.max_leaf_size;
  training_data_ = training_data;

  forest_.resize(header.forest_count);
  nodes_.resize(header.node_count);
  leaf_items_.resize(header.leaf_item_count);
  normals_.resize(header.normal_count);

  read_vector(in, forest_);
  read_vector(in, nodes_);
  read_vector(in, leaf_items_);
  read_vector(in, normals_);
}

std::span<const size_t> AnnoyIndex::query_tree_(const size_t node_idx,
                                                const Vector &query) const {
  DiskNode tree = nodes_[node_idx];
  if (tree.left_index == -1 && tree.right_index == -1) {
    return std::span{
        leaf_items_.data() + tree.leaf_item_offset,
        tree.leaf_item_count,
    };
  }

  const auto normal = std::span{
      normals_.data() + tree.normal_offset,
      dims_,
  };
  const auto is_left = (dot(normal, query) - tree.hyperplane_offset) <= 0;
  if (is_left)
    return query_tree_(tree.left_index, query);
  else
    return query_tree_(tree.right_index, query);
}

std::vector<std::size_t> AnnoyIndex::query(const Vector &query,
                                           std::size_t k) const {
  std::unordered_set<std::size_t> candidate_set;

  for (const auto root_idx : forest_) {
    auto indices = query_tree_(root_idx, query);
    candidate_set.insert(indices.begin(), indices.end());
  }

  std::vector<std::size_t> candidates(candidate_set.begin(),
                                      candidate_set.end());
  std::vector<std::pair<double, std::size_t>> scored_candidates;

  for (const auto c : candidates) {
    scored_candidates.push_back(
        {squared_euclidean_distance(training_data_[c].features, query), c});
  }

  std::partial_sort(scored_candidates.begin(),
                    scored_candidates.begin() +
                        std::min(k, scored_candidates.size()),
                    scored_candidates.end());

  candidates.clear();
  for (const auto &[dist, idx] : scored_candidates) {
    candidates.push_back(idx);
  }

  if (candidates.size() > k) {
    candidates.resize(k);
  }

  return candidates;
}

double AnnoyIndex::evaluate(const Dataset &test_data, std::size_t k) const {
  std::size_t correct = 0;
  const std::size_t progress_every =
      std::max<std::size_t>(1, test_data.size() / 20);

  for (std::size_t i = 0; i < test_data.size(); ++i) {
    const auto &dp = test_data[i];
    const auto neighbors = query(dp.features, k);
    if (classify(neighbors, training_data_) == dp.label) {
      ++correct;
    }

    if ((i + 1) % progress_every == 0 || i + 1 == test_data.size()) {
      std::cout << "  evaluated " << (i + 1) << "/" << test_data.size()
                << " queries...\n";
    }
  }

  return static_cast<double>(correct) / test_data.size();
}

std::size_t AnnoyIndex::n_trees() const { return n_trees_; }

std::size_t AnnoyIndex::max_leaf_size() const { return max_leaf_size_; }

const std::vector<std::size_t> &AnnoyIndex::forest() const { return forest_; }

} // namespace star_nn
